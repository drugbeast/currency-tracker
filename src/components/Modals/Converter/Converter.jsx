import { CURRENCIES_SECOND_FORM, RATES_FOR_TESTS } from 'Constants/constants'
import { useEffect, useState } from 'react'

import ConversionResult from '../../Core/ConversionResult/ConversionResult'
import ErrorBoundary from '../../Core/ErrorBoundary/ErrorBoundary'
import styles from './Converter.module.scss'

function Converter(props) {
  const { currency } = props
  const rates = !window.Cypress
    ? JSON.parse(localStorage.getItem('rates'))
    : RATES_FOR_TESTS

  const [selectedCurrency, setSelectedCurrency] = useState('')
  const [amount, setAmount] = useState(0)
  const [result, setResult] = useState(amount)

  useEffect(() => {
    rates.forEach((item) => {
      if (item.symbol === selectedCurrency) {
        setResult((item.rate * amount) / currency.rate)
      }
    })
  }, [amount, selectedCurrency])

  const handleSetAmount = (e) => {
    setAmount(e.target.value)
  }

  const handleSelectedCurrency = (e) => {
    setSelectedCurrency(e.target.value)
  }

  return (
    <>
      <span className={styles.title}>Converter</span>
      <section className={styles.content} data-cy="converter-modal-wrapper">
        <section className={styles.user}>
          <span className={styles.symbol}>{currency.symbol}:</span>
          <input
            type="number"
            className={styles.amount}
            onChange={handleSetAmount}
            value={amount}
            data-cy="converter-input"
          />
        </section>
        <div className={styles.selectBlock}>
          <span className={styles.to}>to:</span>
          <select
            value={selectedCurrency}
            onChange={handleSelectedCurrency}
            className={styles.select}
            data-cy="converter-select"
          >
            <option value="first">Select, please</option>
            {Object.entries(CURRENCIES_SECOND_FORM).map(
              (name, index) =>
                name[0] !== currency.symbol && (
                  <option value={name[0]} key={index}>
                    {name[0]} ({name[1]})
                  </option>
                ),
            )}
          </select>
        </div>
      </section>
      <ErrorBoundary result={result}>
        <ConversionResult result={result} currency={selectedCurrency} />
      </ErrorBoundary>
    </>
  )
}

export default Converter
