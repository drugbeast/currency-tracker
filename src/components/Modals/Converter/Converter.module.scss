@use '../../../constants/mixins.scss' as *;

.title {
  font-weight: 400;
  font-size: 32px;
  margin: 0 auto;
}

.content {
  margin-top: 48px;

  @include flex(stretch, stretch, 16px, column);
}

.symbol {
  font-weight: 400;
}

.select {
  margin-left: 10%;
  border-radius: 5px;
  padding: 5px;
  width: 80%;
  outline: none;
}

.to {
  font-weight: 300;
}

.amount::-webkit-outer-spin-button,
.edit-input::-webkit-outer-spin-button,
.edit-input::-webkit-inner-spin-button .amount::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}
.amount {
  font-weight: 300;
  outline: none;
  padding: 8px;
  border-radius: 5px;
  border: 0;
  margin-left: 8px;
  width: 80%;
}

@media screen and (max-width: 1024px) {
  .title {
    font-size: 28px;
  }

  .result {
    font-size: 32px;
  }
}
