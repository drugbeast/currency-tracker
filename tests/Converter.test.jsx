import '@testing-library/jest-dom'

import { fireEvent, render, screen } from '@testing-library/react'
import Converter from 'Components/Modals/Converter/Converter'

jest.spyOn(Storage.prototype, 'setItem')

describe('Converter Component', () => {
  const mockCurrency = {
    symbol: 'USD',
    rate: 1.2,
  }
  const mockRates = [{ rate: 1.3, symbol: 'EUR' }]
  const mockConversionResult = '108.333 EUR'
  const mockConversionAmount = '100'
  const mockConversionCurrency = 'EUR'

  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('should render the component', () => {
    localStorage.setItem('rates', JSON.stringify(mockRates))

    render(<Converter currency={mockCurrency} />)
    const titleElement = screen.getByText('Converter')
    const modalWrapperElement = screen.getByTestId('converter-modal-wrapper')
    const inputElement = screen.getByTestId('converter-input')
    const selectElement = screen.getByTestId('converter-select')

    expect(titleElement).toBeInTheDocument()
    expect(modalWrapperElement).toBeInTheDocument()
    expect(inputElement).toBeInTheDocument()
    expect(selectElement).toBeInTheDocument()
  })

  it('should update amount and selected currency on input and select change', () => {
    localStorage.setItem('rates', JSON.stringify(mockRates))

    render(<Converter currency={mockCurrency} />)
    const inputElement = screen.getByTestId('converter-input')
    const selectElement = screen.getByTestId('converter-select')

    fireEvent.change(inputElement, { target: { value: mockConversionAmount } })
    fireEvent.change(selectElement, { target: { value: mockConversionCurrency } })

    expect(inputElement.value).toBe(mockConversionAmount)
    expect(selectElement.value).toBe(mockConversionCurrency)
  })

  it('should calculate and display the conversion result correctly', () => {
    localStorage.setItem('rates', JSON.stringify(mockRates))

    render(<Converter currency={mockCurrency} />)
    const inputElement = screen.getByTestId('converter-input')
    const selectElement = screen.getByTestId('converter-select')
    const conversionResultElement = screen.getByTestId('conversion-result')

    fireEvent.change(inputElement, { target: { value: mockConversionAmount } })
    fireEvent.change(selectElement, { target: { value: mockConversionCurrency } })

    expect(conversionResultElement).toHaveTextContent(mockConversionResult)
  })

  it('should match a snapshot', () => {
    const ConverterComponent = render(<Converter currency={mockCurrency} />)
    expect(ConverterComponent).toMatchSnapshot()
  })
})
