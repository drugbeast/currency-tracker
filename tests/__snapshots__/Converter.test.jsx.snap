// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Converter Component should match a snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <span
        class="title"
      >
        Converter
      </span>
      <section
        class="content"
        data-cy="converter-modal-wrapper"
        data-testid="converter-modal-wrapper"
      >
        <section
          class="user"
        >
          <span
            class="symbol"
          >
            USD
            :
          </span>
          <input
            class="amount"
            data-cy="converter-input"
            data-testid="converter-input"
            type="number"
            value="0"
          />
        </section>
        <div
          class="selectBlock"
        >
          <span
            class="to"
          >
            to:
          </span>
          <select
            class="select"
            data-cy="converter-select"
            data-testid="converter-select"
          >
            <option
              value="first"
            >
              Select, please
            </option>
            <option
              value="AUD"
            >
              AUD
               (
              Australian Dollar
              )
            </option>
            <option
              value="BTC"
            >
              BTC
               (
              Bitcoin
              )
            </option>
            <option
              value="ARS"
            >
              ARS
               (
              Argetine Peso
              )
            </option>
            <option
              value="CAD"
            >
              CAD
               (
              Canadian Dollar
              )
            </option>
            <option
              value="EUR"
            >
              EUR
               (
              Euro
              )
            </option>
            <option
              value="TRY"
            >
              TRY
               (
              Libra
              )
            </option>
            <option
              value="KRW"
            >
              KRW
               (
              Won
              )
            </option>
            <option
              value="JPY"
            >
              JPY
               (
              Yen
              )
            </option>
          </select>
        </div>
      </section>
      <span
        class="result"
        data-cy="conversion-result"
        data-testid="conversion-result"
        role="presentation"
      >
        0
         
      </span>
    </div>
  </body>,
  "container": <div>
    <span
      class="title"
    >
      Converter
    </span>
    <section
      class="content"
      data-cy="converter-modal-wrapper"
      data-testid="converter-modal-wrapper"
    >
      <section
        class="user"
      >
        <span
          class="symbol"
        >
          USD
          :
        </span>
        <input
          class="amount"
          data-cy="converter-input"
          data-testid="converter-input"
          type="number"
          value="0"
        />
      </section>
      <div
        class="selectBlock"
      >
        <span
          class="to"
        >
          to:
        </span>
        <select
          class="select"
          data-cy="converter-select"
          data-testid="converter-select"
        >
          <option
            value="first"
          >
            Select, please
          </option>
          <option
            value="AUD"
          >
            AUD
             (
            Australian Dollar
            )
          </option>
          <option
            value="BTC"
          >
            BTC
             (
            Bitcoin
            )
          </option>
          <option
            value="ARS"
          >
            ARS
             (
            Argetine Peso
            )
          </option>
          <option
            value="CAD"
          >
            CAD
             (
            Canadian Dollar
            )
          </option>
          <option
            value="EUR"
          >
            EUR
             (
            Euro
            )
          </option>
          <option
            value="TRY"
          >
            TRY
             (
            Libra
            )
          </option>
          <option
            value="KRW"
          >
            KRW
             (
            Won
            )
          </option>
          <option
            value="JPY"
          >
            JPY
             (
            Yen
            )
          </option>
        </select>
      </div>
    </section>
    <span
      class="result"
      data-cy="conversion-result"
      data-testid="conversion-result"
      role="presentation"
    >
      0
       
    </span>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
